node/lib/util/checkout.js

```diff
@@ -81,10 +112,51 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
         }
 
         const repo = yield SubmoduleUtil.getRepo(metaRepo, name);
+        subRepos.push(repo);
         const sha = shas[name];
         yield GitUtil.fetchSha(repo, sha);
         const commit = yield repo.getCommit(sha);
-        yield GitUtil.setHeadHard(repo, commit);
+        subCommits.push(commit);
+        const error = yield dryRun(repo, commit);
+        if (null !== error) {
+            errors.push(
+             `Unable to checkout submodule ${colors.yellow(name)}: ${error}.`);
+        }
+    }));
+
+    // Throw an error if any dry-runs failed.
+
+    if (0 !== errors.length) {
+        throw new UserError(errors.join("\n"));
+    }
+
+    /**
+     * Checkout and set as head the specified `commit` in the specified `repo`.
+     */
+    const doCheckout = co.wrap(function *(repo, commit) {
+        yield NodeGit.Checkout.tree(repo, commit, {
```

Comment:
The only way I could make it behave like `git checkout` was to run it twice; basically, once to apply to the index and once to the workdir. I'll add a comment to explain.